src/main/java/io/strimzi/CertificateWatch.java

```diff
@@ -0,0 +1,266 @@
+     * the secret is deleted, we do not do anything -> in such case, the new Pod would not start anyway. We expect the
+     * secret to be created again later which is when we might restart if the certificates differ.
+     */
+    private void setupInformer()    {
+        secretInformer = client.secrets().inNamespace(namespace).withName(secretName).inform(new ResourceEventHandler<>() {
+            @Override
+            public void onAdd(Secret secret) {
+                LOG.info("Secret {} was added and will be checked for changes", secret.getMetadata().getName());
+                checkForChanges(secret);
+            }
+
+            @Override
+            public void onUpdate(Secret oldSecret, Secret newSecret) {
+                LOG.info("Secret {} was modified and will be checked for changes", newSecret.getMetadata().getName());
+                checkForChanges(newSecret);
+            }
+
+            @Override
+            public void onDelete(Secret secret, boolean deletedFinalStateUnknown) {
+                LOG.info("Secret {} was deleted -> nothing to do", secret.getMetadata().getName());
+            }
+        });
+    }
+
+    /**
+     * Checks the Secret for changes. It extracts the selected keys from it and compares it to the previously known
+     * value. If they differ, it triggers the restart of the Drain Cleaner.
+     *
+     * @param secret    Secret with the values which should be compared with the known certificates
+     */
+    /* test */ void checkForChanges(Secret secret) {
+        if (secret == null
+                || !previousValues.equals(getSecretData(secret))) {
+            LOG.info("The watched fields {} changed and Drain Cleaner restart will be triggered", secretKeys);
+
+            // The pod is deleted asynchronously to not block the event handler from the secretInformer
+            // Doing it synchronously was causing Fabric8 exceptions since it was blocking the thread while the informer was
+            // shutting down => because it is deleting itself
+            CompletableFuture.supplyAsync(() -> {
+                restartDrainCleaner();
+                return (Void) null;
+            });
+        } else {
+            LOG.info("No change to watched fields ({}) detected", secretKeys);
+        }
+    }
+
+    /**
+     * Restarts this Drain Cleaner Pod by deleting it through the Kubernetes API
+     */
+    /* test */ void restartDrainCleaner()  {
+        LOG.info("Deleting pod {} to restart Drain Cleaner and reload certificates", podName);
+        client.pods().inNamespace(namespace).withName(podName).delete();
+    }
+
+    /**
+     * Extracts the data from the Secret
+     *
+     * @param secret   Secret from which the data should be extracted
+     *
+     * @return  The values of selected keys from the Secret as one long String
+     */
+    private String getSecretData(Secret secret) {
+        StringBuilder sb = new StringBuilder();
+
+        if (secret.getData() != null) {
```

Comment:
Do we know that the `data` an ordered map? If it's not ordered we can't rely on the equality of the returned String to detect a change. We might want to cope the `data` into a `TreeMap` to give it a canonical order.

Reply:
I don't think the ordering in the map matters. We look up the keys from the list and the list is created in the constructor. So even if the ordering of the list is not guaranteed, I would think that it doesn't matter since it will be the same in the whole runtime of the Drain Cleaner pod and change only when it restarts (when it anyway loads the new initial state?)

Reply:
I did some googling and as far as I found, the iteration over a list should have a consistent order. So I think we should be fine here since every call will iterate through the keys in the same order and make sure the String is correct regardless of the Secret map. Please let me know if it makes sense or if I missed something.

Reply:
I think that's correct from the Java PoV: So long as the same keys are added (to a new HashMap, or LinkedHashMap, or TreeMap) in the same order then they'd have the same iteration order, at least within that JVM). And we would expect the order to be stable between observations of an unmodified resource (because we expect that the JSON stored in etcd isn't getting changed). So I can see that this _works_. OTOH I don't think we can point to actual documentation (contracts) that proves it would always work. For example, if something touched the Secret and ended up reordering the keys, even if the mapping were unchanged then it would fail. We expect that to be unlikely, and a false positive is not the end of the world, so I guess this is all fine. OTOH is seems trivial to iterate over `new TreeMap(secret.getData())` rather than just `secret.getData()`, which would make it bullet proof.

I guess I'm fine either way.

Reply:
But we do not iterate the Map from the Secret. If we did that, I do not think the order will be guaranteed. The Drain Cleaner configuration has a list of keys from the secret we are interested in. And this list is what we iterate through -> and for each key we just call `get(...)` on the map.